### PR TITLE
MODORDERS-178/179 Increased flexibility in inventory integration

### PR DESCRIPTION
**Pull request by grabsefx**
## Purpose
https://issues.folio.org/browse/MODORDERS-178
https://issues.folio.org/browse/MODORDERS-179

## Approach
* Updating `physical.json` and `eresource.json` schemas by updating acq-models submodule
* Implement creation of inventory objects based on resource createInventory values
* Implementing loading default createInventory values from mod-configuration. In case of empty configuration set hardcoded values (MODORDERS-179)
* Implementing unit tests for both stories

#### TODOS and Open Questions
* As for PR contains breaking changes, firstly PR https://github.com/folio-org/acq-models/pull/102 should be merged and acq-model submodule updated. Also PR for the story https://issues.folio.org/browse/UIOR-148 should be merged in the same day.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x ] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [x] Were any API paths or methods changed, added or removed?
  - [x] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/orders/utils/HelperUtils.java [hunk 1]
```diff
@@ -776,4 +783,25 @@ public static int getPoLineLimit(JsonObject config) {
   public static CompositePurchaseOrder convertToCompositePurchaseOrder(JsonObject poJson) {
     return poJson.mapTo(CompositePurchaseOrder.class);
   }
+
+  public static boolean inventoryUpdateNotRequired(CompositePoLine compPOL) {
+    return (compPOL.getPhysical() == null && compPOL.getEresource() == null) ||
+      (compPOL.getPhysical() != null && compPOL.getPhysical().getCreateInventory() == Physical.CreateInventory.NONE && compPOL.getEresource() == null) ||
+      (compPOL.getEresource() != null && compPOL.getEresource().getCreateInventory() == Eresource.CreateInventory.NONE && compPOL.getPhysical() == null) ||
+      (compPOL.getPhysical() != null && compPOL.getEresource() != null && compPOL.getEresource().getCreateInventory() == Eresource.CreateInventory.NONE && compPOL.getPhysical().getCreateInventory() == Physical.CreateInventory.NONE) ||
+      // also ignore inventory interaction in case of "Other" order format
```
Comment (craigmcnally) [inline src/main/java/org/folio/orders/utils/HelperUtils.java]: I think there's a misunderstanding here... We don't want to ignore inventory interaction for "Other" orders.  

In the case of "Other", use createInventory and quantity from poLine.physical.  The JIRA describes that the UI should set the default value of physical.createInventory to 'None', but there's nothing preventing a user from changing the drop-down value.

File: src/main/java/org/folio/orders/utils/HelperUtils.java [hunk 2]
```diff
@@ -499,12 +494,14 @@ public static int calculateInventoryItemsQuantity(CompositePoLine compPOL, List<
...
+        quantity += isItemsUpdateRequiredForPhysical(compPOL) ? getPhysicalQuantity(locations) : 0;
+        quantity += isItemsUpdateRequiredForEresource(compPOL) ? getElectronicQuantity(locations) : 0;
+        return quantity;
       case PHYSICAL_RESOURCE:
-        return getPhysicalQuantity(locations);
+        return isItemsUpdateRequiredForPhysical(compPOL) ? getPhysicalQuantity(locations) : 0;
       case ELECTRONIC_RESOURCE:
-        return isInventoryUpdateRequiredForEresource(compPOL) ? getElectronicQuantity(locations) : 0;
+        return isItemsUpdateRequiredForEresource(compPOL) ? getElectronicQuantity(locations) : 0;
       case OTHER:
```
Comment (craigmcnally) [inline src/main/java/org/folio/orders/utils/HelperUtils.java]: See my comment below about inventory interactions for "Other" order lines

File: src/main/java/org/folio/orders/utils/HelperUtils.java [hunk 3]
```diff
@@ -776,4 +783,25 @@ public static int getPoLineLimit(JsonObject config) {
   public static CompositePurchaseOrder convertToCompositePurchaseOrder(JsonObject poJson) {
     return poJson.mapTo(CompositePurchaseOrder.class);
   }
+
+  public static boolean inventoryUpdateNotRequired(CompositePoLine compPOL) {
```
Comment (craigmcnally) [inline src/main/java/org/folio/orders/utils/HelperUtils.java]: This conditional is quite large... We might consider breaking it up, or at least adding additional comments

File: src/main/java/org/folio/orders/utils/HelperUtils.java [hunk 4]
```diff
@@ -776,4 +783,25 @@ public static int getPoLineLimit(JsonObject config) {
...
+    return (compPOL.getPhysical() == null && compPOL.getEresource() == null) ||
+      (compPOL.getPhysical() != null && compPOL.getPhysical().getCreateInventory() == Physical.CreateInventory.NONE && compPOL.getEresource() == null) ||
+      (compPOL.getEresource() != null && compPOL.getEresource().getCreateInventory() == Eresource.CreateInventory.NONE && compPOL.getPhysical() == null) ||
+      (compPOL.getPhysical() != null && compPOL.getEresource() != null && compPOL.getEresource().getCreateInventory() == Eresource.CreateInventory.NONE && compPOL.getPhysical().getCreateInventory() == Physical.CreateInventory.NONE) ||
+      // also ignore inventory interaction in case of "Other" order format
+      compPOL.getOrderFormat() == CompositePoLine.OrderFormat.OTHER;
+  }
+
+  public static boolean isHoldingsUpdateRequired(CompositePoLine compPOL) {
+    return (compPOL.getEresource() != null &&
```
Comment (craigmcnally) [inline src/main/java/org/folio/orders/utils/HelperUtils.java]: This conditional is quite large... We might consider breaking it up, or at least adding additional comments

File: src/main/java/org/folio/orders/utils/HelperUtils.java [hunk 5]
```diff
@@ -776,4 +783,25 @@ public static int getPoLineLimit(JsonObject config) {
...
+      (compPOL.getEresource().getCreateInventory() == Eresource.CreateInventory.INSTANCE_HOLDING || compPOL.getEresource().getCreateInventory() == Eresource.CreateInventory.INSTANCE_HOLDING_ITEM))
+      || (compPOL.getPhysical() != null &&
+      (compPOL.getPhysical().getCreateInventory() == Physical.CreateInventory.INSTANCE_HOLDING || compPOL.getPhysical().getCreateInventory() == Physical.CreateInventory.INSTANCE_HOLDING_ITEM));
+  }
+
+  public static boolean isItemsUpdateRequired(CompositePoLine compPOL) {
+    return (compPOL.getEresource() != null && compPOL.getEresource().getCreateInventory() == Eresource.CreateInventory.INSTANCE_HOLDING_ITEM)
```
Comment (craigmcnally) [inline src/main/java/org/folio/orders/utils/HelperUtils.java]: This conditional is quite large... We might consider breaking it up, or at least adding additional comments

File: src/main/java/org/folio/rest/impl/InventoryHelper.java [hunk 6]
```diff
@@ -103,24 +97,24 @@
...
+        .forEach((locationId, locations) -> itemsPerHolding.add(
+          // Search for or create a new holding and then create items for this holding
+          getOrCreateHoldingsRecord(compPOL, locationId)
+            .thenCompose(holdingId -> {
+              int expectedQuantity = calculateInventoryItemsQuantity(compPOL, locations);
+              if (isItemsUpdateRequired(compPOL) && expectedQuantity > 0) {
+                  // For some cases items might not be created e.g. resources with create inventory set to "None"
+                  return handleItemRecords(compPOL, holdingId, expectedQuantity)
+                    .thenApply(itemIds -> constructPieces(itemIds, compPOL.getId(), locationId));
+                } else return completedFuture((new ArrayList<>()));
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/InventoryHelper.java]: Please use curly bracket notation, even for one-liners like this.  It improves readability

Comment (craigmcnally) [reply src/main/java/org/folio/rest/impl/InventoryHelper.java]: also you indentation might be a little off here, can't tell in github

File: src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java [hunk 7]
```diff
@@ -160,11 +136,41 @@ private CompositePurchaseOrder validateOrderState(CompositePurchaseOrder po) {
...
+      return future
+        .thenAccept(jsonConfig -> updateCreateInventory(compPOL, jsonConfig));
+    } else return completedFuture(null);
+  }
+
+  private void updateCreateInventory(CompositePoLine compPOL, JsonObject jsonConfig) {
+    //try to set createInventory by values from mod-configuration. If empty - set default hardcoded values
+    if (compPOL.getEresource() != null && compPOL.getEresource().getCreateInventory() == null) {
+      String createInventory = jsonConfig.getString(ERESOURCE);
+      Eresource.CreateInventory eresource = StringUtils.isEmpty(createInventory) ?
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java]: this is a bit of a nitpick but this variable name is poorly named... consider removing this temporary var altogether or renaming it to something like `effectiveValue`

File: src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java [hunk 8]
```diff
@@ -160,11 +136,41 @@ private CompositePurchaseOrder validateOrderState(CompositePurchaseOrder po) {
...
+        .exceptionally(t -> future.complete(new JsonObject()));
+      return future
+        .thenAccept(jsonConfig -> updateCreateInventory(compPOL, jsonConfig));
+    } else return completedFuture(null);
+  }
+
+  private void updateCreateInventory(CompositePoLine compPOL, JsonObject jsonConfig) {
+    //try to set createInventory by values from mod-configuration. If empty - set default hardcoded values
+    if (compPOL.getEresource() != null && compPOL.getEresource().getCreateInventory() == null) {
+      String createInventory = jsonConfig.getString(ERESOURCE);
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java]: I think a better variable name might be `tenantDefault`

File: src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java [hunk 9]
```diff
@@ -160,11 +136,41 @@ private CompositePurchaseOrder validateOrderState(CompositePurchaseOrder po) {
...
+      Eresource.CreateInventory eresource = StringUtils.isEmpty(createInventory) ?
+        Eresource.CreateInventory.INSTANCE_HOLDING : Eresource.CreateInventory.fromValue(createInventory);
+      compPOL.getEresource().setCreateInventory(eresource);
+    }
+    if (compPOL.getPhysical() != null && compPOL.getPhysical().getCreateInventory() == null) {
+      String createInventory = jsonConfig.getString(PHYSICAL);
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java]: see notes above about var names

File: src/test/java/org/folio/rest/impl/OrdersImplTest.java [hunk 10]
```diff
@@ -3463,8 +3594,10 @@ private void handleGetAccessProviders(RoutingContext ctx) {
         } else if(getQuery(INACTIVE_ACCESS_PROVIDER_A, INACTIVE_ACCESS_PROVIDER_B).equals(query)
```
Comment (craigmcnally) [inline src/test/java/org/folio/rest/impl/OrdersImplTest.java]: whitespace after `else if`

File: src/test/resources/mockdata/compositeOrders/e5ae4afd-3fa9-494e-a972-f541df9b877e.json [hunk 11]
```diff
@@ -87,6 +87,12 @@
       "orderFormat": "Electronic Resource",
       "owner": "LTS E-Resources & Serials",
       "paymentStatus": "Awaiting Payment",
+      "physical": {
+        "createInventory": "Instance, Holding, Item",
+        "materialSupplier": "36dd6d62-c55e-4b19-a3a0-a870da57610a",
+        "receiptDue": null,
```
Comment (craigmcnally) [inline src/test/resources/mockdata/compositeOrders/e5ae4afd-3fa9-494e-a972-f541df9b877e.json]: Should probably just omit null fields altogether

File: src/main/java/org/folio/orders/utils/HelperUtils.java (hunk 1 above)
Comment (craigmcnally) [reply src/main/java/org/folio/orders/utils/HelperUtils.java]: Please make sure test cases are updated to cover this too

File: src/test/java/org/folio/rest/impl/OrdersImplTest.java (hunk 10 above)
Comment (grabsefx) [reply src/test/java/org/folio/rest/impl/OrdersImplTest.java]: fixed for a whole block of code

File: src/test/resources/mockdata/compositeOrders/e5ae4afd-3fa9-494e-a972-f541df9b877e.json (hunk 11 above)
Comment (grabsefx) [reply src/test/resources/mockdata/compositeOrders/e5ae4afd-3fa9-494e-a972-f541df9b877e.json]: updated with valid value

File: src/main/java/org/folio/rest/impl/InventoryHelper.java (hunk 6 above)
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/InventoryHelper.java]: added brackets. formatted code

File: src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java (hunk 8 above)
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java]: variable naming updated

File: src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java (hunk 7 above)
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java]: variable naming updated

File: src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java (hunk 9 above)
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java]: variable naming updated

File: src/main/java/org/folio/orders/utils/HelperUtils.java (hunk 4 above)
Comment (grabsefx) [reply src/main/java/org/folio/orders/utils/HelperUtils.java]: updated conditional code for better readability

File: src/main/java/org/folio/orders/utils/HelperUtils.java (hunk 5 above)
Comment (grabsefx) [reply src/main/java/org/folio/orders/utils/HelperUtils.java]: updated conditional code for better readability

File: src/main/java/org/folio/orders/utils/HelperUtils.java (hunk 1 above)
Comment (grabsefx) [reply src/main/java/org/folio/orders/utils/HelperUtils.java]: Updated logic for "Other" format.

File: src/main/java/org/folio/orders/utils/HelperUtils.java (hunk 2 above)
Comment (grabsefx) [reply src/main/java/org/folio/orders/utils/HelperUtils.java]: updated items count for "Other" basing on physical quantity.

File: src/main/java/org/folio/orders/utils/HelperUtils.java (hunk 3 above)
Comment (grabsefx) [reply src/main/java/org/folio/orders/utils/HelperUtils.java]: conditions updated for better readability

File: src/main/java/org/folio/rest/impl/InventoryHelper.java [hunk 12]
```diff
@@ -103,24 +97,26 @@
    * @param compPOL   PO line to retrieve/create Item Records for. At this step PO Line must contain instance Id
    * @return future with list of pieces with item and location id's
    */
-  public CompletableFuture<List<Piece>> handleItemRecords(CompositePoLine compPOL) {
+  public CompletableFuture<List<Piece>> handleHoldingRecords(CompositePoLine compPOL) {
```
[the remark this answers is not in the document]
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/InventoryHelper.java]: naming updated

File: src/main/java/org/folio/rest/impl/InventoryHelper.java [hunk 13]
```diff
@@ -103,24 +97,26 @@
...
+          // Search for or create a new holding and then create items for this holding
+          getOrCreateHoldingsRecord(compPOL, locationId)
+            .thenCompose(holdingId -> {
+                int expectedQuantity = calculateInventoryItemsQuantity(compPOL, locations);
+                if (isItemsUpdateRequired(compPOL) && expectedQuantity > 0) {
+                  // For some cases items might not be created e.g. resources with create inventory set to "None"
+                  return handleItemRecords(compPOL, holdingId, expectedQuantity)
+                    .thenApply(itemIds -> constructPieces(itemIds, compPOL.getId(), locationId));
+                } else {
+                  return completedFuture((new ArrayList<>()));
```
[the remark this answers is not in the document]
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/InventoryHelper.java]: fixed

File: src/main/java/org/folio/orders/utils/HelperUtils.java [hunk 14]
```diff
@@ -779,4 +785,43 @@ public static int getPoLineLimit(JsonObject config) {
...
+        || compPOL.getEresource().getCreateInventory() == Eresource.CreateInventory.INSTANCE_HOLDING_ITEM);
+    }
+    if (compPOL.getPhysical() != null) {
+      updateRequiredForPhysical = (compPOL.getPhysical().getCreateInventory() == Physical.CreateInventory.INSTANCE_HOLDING
+        || compPOL.getPhysical().getCreateInventory() == Physical.CreateInventory.INSTANCE_HOLDING_ITEM);
+    }
+    return updatesRequiredForEresource || updateRequiredForPhysical;
+  }
+
+  public static boolean isItemsUpdateRequired(CompositePoLine compPOL) {
```
[the remark this answers is not in the document]
Comment (grabsefx) [reply src/main/java/org/folio/orders/utils/HelperUtils.java]: fixed

File: src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java [hunk 15]
```diff
@@ -246,22 +255,21 @@ private CompositePurchaseOrder validateOrderState(CompositePurchaseOrder po) {
     if (compPOL.getReceiptStatus() == CompositePoLine.ReceiptStatus.RECEIPT_NOT_REQUIRED) {
       return completedFuture(null);
     }
+    // Set InstanceId only for the checkin flow
     if (compPOL.getCheckinItems() != null && compPOL.getCheckinItems()) {
```
[the remark this answers is not in the document]
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java]: fixed

File: src/main/java/org/folio/rest/impl/InventoryHelper.java [hunk 16]
```diff
@@ -103,24 +97,26 @@
...
-      });
-
+    if (HelperUtils.isHoldingsUpdateRequired(compPOL)) {
+      groupLocationsById(compPOL)
+        .forEach((locationId, locations) -> itemsPerHolding.add(
+          // Search for or create a new holding and then create items for this holding
+          getOrCreateHoldingsRecord(compPOL, locationId)
+            .thenCompose(holdingId -> {
+                int expectedQuantity = calculateInventoryItemsQuantity(compPOL, locations);
+                if (isItemsUpdateRequired(compPOL) && expectedQuantity > 0) {
```
[the remark this answers is not in the document]
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/InventoryHelper.java]: fixed

File: src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java [hunk 17]
```diff
@@ -160,11 +136,43 @@ private CompositePurchaseOrder validateOrderState(CompositePurchaseOrder po) {
...
       .thenAccept(lineNumber -> line.put(PO_LINE_NUMBER, lineNumber))
       .thenCompose(v -> createPoLineSummary(compPoLine, line));
   }
 
+  CompletableFuture<Void> setTenantDefaultCreateInventoryValues(CompositePoLine compPOL) {
+    CompletableFuture<JsonObject> future = new VertxCompletableFuture<>(ctx);
+
+    if ((compPOL.getPhysical() != null && compPOL.getPhysical().getCreateInventory() == null)
+      || (compPOL.getEresource() != null && compPOL.getEresource().getCreateInventory() == null)) {
+      loadConfiguration(okapiHeaders, ctx, logger)
```
[the remark this answers is not in the document]
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java]: created method `getTenantConfiguration` in `AbstractHelper`

File: src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java [hunk 18]
```diff
@@ -160,11 +136,43 @@ private CompositePurchaseOrder validateOrderState(CompositePurchaseOrder po) {
...
+        .thenApply(config -> future.complete(Optional.of(config.getJsonObject(CREATE_INVENTORY)).orElse(new JsonObject())))
```
[the remark this answers is not in the document]
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/PurchaseOrderLineHelper.java]: fixed

**Pull request merged**